### ISSUE TITLE
Question regarding self-critical reward 

### ISSUE THREAD

**Issue by fawazsammani** [ruotianluo/self-critical.pytorch#138]
Hi Ruotian,
I'd like to ask about the nature of the reward when training with self-critical. Is it normal to start at negative or zero for the first few epochs? 
I am getting the following for the first epoch with self-critical (after training with XE for 13 epochs):

![Capture](https://user-images.githubusercontent.com/30661597/67779079-1c1e4280-fa9f-11e9-9b84-7857625c6102.PNG)

Is this normal? And what is the maximum reward you achieved after training with self-critical?

Also, I'm using the py3 branch. I saw there are a lot of differences between the py3 and py2 branches. So is the py3 branch reliable to use?

Looking forward to your answer!

**Comment by ruotianluo** [comment]
I think py3 branch is already merged into master. You might Be able to run master branch under py3 if I remember correctly. 


The reward is a little bit weird, because it's decreasing. And, Usually I would expect the initial avg reward should be negative and then start to increase.

**Comment by fawazsammani** [comment]
Thanks for your reply @ruotianluo. I've spotted all the differences between py2 and py3. Is it possible to tell me if they affect anything in self-critical training? I really appreciate your kind help. 

**In py2 but not py3:**
in `preprocess ngrams.py`

```
# Load bpe
  if 'bpe' in dict_json:
    import tempfile
    import codecs
    codes_f = tempfile.NamedTemporaryFile(delete=False)
    codes_f.close()
    with open(codes_f.name, 'w') as f:
      f.write(dict_json['bpe'])
    with codecs.open(codes_f.name, encoding='UTF-8') as codes:
      bpe = apply_bpe.BPE(codes)
    params.bpe = bpe
```

**In py2 but not py3:**
in `LossWrapper` forward:
```
gts = [gts[_] for _ in gt_indices.tolist()]
reward = get_self_critical_reward(greedy_res, gts, gen_result, self.opt)
```

where 
`gts = data['gts']` and `gt_indices = torch.arange(0, len(data['gts'])`

**Difference between py2 and py3:**

In `sample_next_word(self, logprobs, sample_method, temperature):`

`sample_method = 'sample'` when sampling (not greedy result) for self-critical.
So this segment of the code will run (since sample_method does not start with `"top"`:

```
it = torch.distributions.Categorical(logits=logprobs.detach()).sample()
sampleLogprobs = logprobs.gather(1, it.unsqueeze(1)) # gather the logprobs at sampled positions
```

But in py3 version, you have:

```
prob_prev = torch.exp(logprobs)     # fetch prev distribution (softmax)
it = torch.multinomial(prob_prev, 1)
sampleLogprobs = logprobs.gather(1, it) # gather the logprobs at sampled positions
it = it.view(-1).long() # flatten indices for saving in tensor
```

I look forward to your kind reply! It would really help me spot my problem. Thanks alot

**Comment by ruotianluo** [comment]
I think these changes are changed after the merge. Can you verify me if py3 commit is a ancestor of current master branch?

**Comment by fawazsammani** [comment]
It is merged. I didn't clone it that's why it appears different. So these changes above do need to be made if i'm using the parent py3 (not merged) version right?

**Comment by ruotianluo** [comment]
It is fine not to update. These code shouldn't affect training.

**Comment by fawazsammani** [comment]
Alright. I'll check the problem in my code. Thanks a lot for your reply

**Comment by fawazsammani** [comment]
Hello @ruotianluo 
I've managed to solve the problem. Now, the cider is increasing, but the reward is still negative (4 epochs till now). Is that normal to be negative for the first few epochs? And what is the maximum reward you achieved? 

![1](https://user-images.githubusercontent.com/30661597/67851359-44607c80-fb45-11e9-9174-5c592681963d.PNG)

**Comment by ruotianluo** [comment]
It's totally normal! Don't worry.

**Comment by fawazsammani** [comment]
Thanks a lot! Appreciate your kind help
